File: src/main/scala/discard/DiscardClient.scala
package discard

import io.netty.bootstrap.Bootstrap
import io.netty.channel.ChannelInitializer
import io.netty.channel.nio.NioEventLoopGroup
import io.netty.channel.socket.SocketChannel
import io.netty.channel.socket.nio.NioSocketChannel
import io.netty.handler.ssl.SslContextBuilder
import io.netty.handler.ssl.util.InsecureTrustManagerFactory

object DiscardClient {
  private val ssl = true
  private val host = "127.0.0.1"
  private val port = 8080
  private val workGroup = new NioEventLoopGroup
  private val bootstrap = new Bootstrap
  private val sslContext = if (ssl) SslContextBuilder.forClient().trustManager(InsecureTrustManagerFactory.INSTANCE).build() else null
  bootstrap
    .group(workGroup)
    .channel(classOf[NioSocketChannel])
    .remoteAddress(host, port)
    .handler(new ChannelInitializer[SocketChannel] {
      override def initChannel(ch: SocketChannel): Unit = {
        val pipeline = ch.pipeline()
        if (sslContext != null) {
          pipeline.addLast(sslContext.newHandler(ch.alloc(), host, port))
        }
        pipeline.addLast(new DiscardClientHandler)
      }
    })

  def main(args: Array[String]): Unit = {
    try {
      bootstrap
        .connect()
        .sync()
        .channel()
        .closeFuture()
        .sync()
    } finally {
      workGroup.shutdownGracefully()
    }
  }
}

File: src/main/kotlin/com/mdreamfever/netty/example/discard/DiscardClient.kt
package com.mdreamfever.netty.example.discard

import io.netty.bootstrap.Bootstrap
import io.netty.channel.ChannelInitializer
import io.netty.channel.nio.NioEventLoopGroup
import io.netty.channel.socket.SocketChannel
import io.netty.channel.socket.nio.NioSocketChannel
import io.netty.handler.ssl.SslContextBuilder
import io.netty.handler.ssl.util.InsecureTrustManagerFactory
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext

fun main() = runBlocking {
    withContext(Dispatchers.IO) {
        DiscardClient().connect()
    }
}

class DiscardClient(
    private val ssl: Boolean = System.getProperty("ssl") != null,
    private val host: String = System.getProperty("host", "127.0.0.1"),
    private val port: Int = System.getProperty("port", "8009").toInt(),
    val size: Int = System.getProperty("size", "256").toInt(),
) {
    private val bootstraps: Bootstrap = Bootstrap()
    private val group: NioEventLoopGroup = NioEventLoopGroup()

    init {
        val sslContext =
            if (ssl) SslContextBuilder.forClient().trustManager(InsecureTrustManagerFactory.INSTANCE).build() else null
        bootstraps.group(group).channel(NioSocketChannel::class.java)
            .handler(object : ChannelInitializer<SocketChannel>() {
                override fun initChannel(ch: SocketChannel) {
                    val pipeline = ch.pipeline()
                    sslContext?.let {
                        pipeline.addLast(it.newHandler(ch.alloc(), host, port))
                    }
                    pipeline.addLast(DiscardClientHandler())
                }
            })
    }

    fun connect() {
        try {
            val future = bootstraps.connect(host, port).sync()
            future.channel().closeFuture().sync()
        } finally {
            group.shutdownGracefully()
        }
    }
}



